tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/lhlo_legalize_roots_to_loops.cc

```diff
@@ -0,0 +1,329 @@
+      return failure();
+    }
+  }
+  // remove the root_op if it has no other users except the memref
+  if (non_fusion) {
+    for (auto* root_op : root_ops) {
+      root_op->erase();
+    }
+  } else {
+    assert(parent != nullptr && "Parent must be provided for fusion lowering");
+    cleanUnusedLhloOps(parent);
+  }
+  b.create<scf::YieldOp>(loc, ValueRange({}));
+  return success();
+}
+
+bool isOnGpu(Operation* op) {
+  if (isa<FusionOp>(op)) {
+    // TODO(disc): Revisit this when fusion on cpu is suppported
+    return true;
+  }
+  assert(isa<LmhloOp>(op) && "Unexpected usage of isOnGpu");
+  auto result_memref = cast<LmhloOp>(op).getResultBuffer();
+  auto memory_space =
+      result_memref.getType().cast<MemRefType>().getMemorySpace();
+  return memory_space && memory_space.isa<StringAttr>() &&
+         memory_space.cast<StringAttr>().getValue() == placement_utils::c_gpu;
+}
+
+}  // namespace
+
+// Expand the root ops in a fused func into a set of nested loops.
+// This pass must be executed after the fusion pass, and works together with
+// the InputInlineFusion pass after it for fusion codegen.
+//
+// TODO(disc): Currently this pass supports lmhlo.FusionOp to have lmhlo ops
+// inside, not mhlo. It's mainly because we now do fusion on lmhlo, not mhlo.
+// The fusion pass can be moved to mhlo after shape dialect is brought in to
+// represent shape calculation on tensor layer, and we would be able to do shape
+// calculation lowering for mhlo.FusionOp. Reconsider the fusion representation
+// after these are done, a lmhlo.FusionOp with mhlo inside would be more
+// friendly to the legacy FusedIrEmitter.
+class LhloLegalizeRootsToParallelLoops
+    : public LhloLegalizeRootsToParallelLoopsPassBase<
+          LhloLegalizeRootsToParallelLoops> {
+  void runOnFunction() override {
+    auto func = getFunction();
+    OpBuilder b(func);
+    SmallVector<Operation*, 4> gpu_non_fusion_worklist;
+    SmallVector<Operation*, 4> cpu_non_fusion_worklist;
+    SmallVector<Operation*, 4> gpu_fusion_worklist;
+    func.walk([&](LmhloOp op) {
+      // Skip nested computation of reduce etc.
+      if (op->getParentOp() != func.getOperation()) {
+        return;
+      }
+      if (isa<FusionOp>(op)) {
+        // TODO(disc): Revisit this when fusion on cpu is supported
+        gpu_fusion_worklist.push_back(op);
+      } else if (isOnGpu(op)) {
+        gpu_non_fusion_worklist.push_back(op);
+      } else {
+        cpu_non_fusion_worklist.push_back(op);
+      }
+    });
+
+    for (auto op : cpu_non_fusion_worklist) {
+      // Only for calculating shapes when the backend is gpu. A simple schedule
+      // should be sufficient for performance.
+      // TODO(disc): Revisit this when the backend is cpu and the calculation is
+      // for data
+      auto r =
+          lowerWithScheduleLoop(b, ArrayRef<Operation*>{op}, op, nullptr,
+                                /*non_fusion*/ true, /*parallel_loop*/ false);
```

Comment:
```suggestion
                                /*non_fusion=*/ true, /*parallel_loop=*/ false);
```

This annotation format is recognized by clang-tidy, which warns on mismatch with the function invoked.